### PR TITLE
[FLINK-19072][table-planner] Import Temporal Table join rule for Stream

### PR DESCRIPTION
**Pull request by leonardBang**
## What is the purpose of the change

* This pull request supports Temporal table join in planner.

## Brief change log

  - ea54f7198e : [FLINK-19072] import Temporal Table join rule
(1) Import temporal table join rule to match all temporal join, lookup join also belong temporal join 
(2) Defferentiate lookup join and general temporal join in according the join pattern
 
 - 61ab18c9e4012d0c65d64c2456af0900577b90db : [FLINK-19073] Imporve streamExecTemporalJoinRule
(1) Match temporal table join from temporal function in StreamExecLegacyTemporalJoinRule,  Match temporal table join in StreamTemporalJoinRule
(2) Infer changelog mode trait for temporal table join node

  - 1b5a16ff124bb762352c8f31305ea3d3fcf4b694 : [FLINK-19076] Import rule to deal Temporal Join condition
(1) Obtains the unique key from versioned table/view
(2) Prevent join condition(e.g. time attribute column, primary key) push down by adding temporal join condition in join node

## Verifying this change

Add plan tests  to cover existed lookup join, temporal function and temporal table join

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
JavaDocs, will add docs in FLINK-19082

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/StreamExecTemporalJoinRule.scala [hunk 1]
```diff
@@ -18,56 +18,53 @@
...
-import org.apache.flink.table.planner.plan.utils.{FlinkRelOptUtil, IntervalJoinUtil}
+import java.util
 
 import org.apache.calcite.plan.RelOptRule.{any, operand}
 import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall, RelTraitSet}
 import org.apache.calcite.rel.RelNode
 import org.apache.calcite.rel.core.JoinRelType
 
-import java.util
+import org.apache.flink.table.planner.plan.`trait`.FlinkRelDistribution
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/StreamExecTemporalJoinRule.scala:28]: reorder imports

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/LegacyTemporalProcessTimeJoinOperator.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.util.RowDataUtil;
+import org.apache.flink.table.runtime.generated.GeneratedJoinCondition;
+import org.apache.flink.table.runtime.generated.JoinCondition;
+import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
+
+/**
+ * The operator to temporal join a stream on processing time.
+ */
+public class LegacyTemporalProcessTimeJoinOperator
```
Comment (godfreyhe) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/LegacyTemporalProcessTimeJoinOperator.java:40]: `LegacyTemporalProcessTimeJoinOperator` and `TemporalProcessTimeJoinOperator` is almost same, can we reuse them ?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/LegacyTemporalRowTimeJoinOperator.java [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+ * by registering timers for the keys. We could register a timer for every probe and build
+ * side element's event time (when watermark exceeds this timer, that's when we are emitting and/or
+ * cleaning up the state). However this would cause huge number of registered timers. For example
+ * with following evenTimes of probe records accumulated: {1, 2, 5, 8, 9}, if we
+ * had received Watermark(10), it would trigger 5 separate timers for the same key. To avoid that
+ * we always keep only one single registered timer for any given key, registered for the minimal
+ * value. Upon triggering it, we process all records with event times older then or equal to
+ * currentWatermark.
+ */
+public class LegacyTemporalRowTimeJoinOperator
```
Comment (godfreyhe) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/LegacyTemporalRowTimeJoinOperator.java:72]: ditto

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala [hunk 4]
```diff
@@ -42,11 +41,11 @@ import org.apache.calcite.rex.{RexCall, RexInputRef, RexNode}
 import org.apache.calcite.sql.SqlKind
 import org.apache.calcite.sql.fun.SqlStdOperatorTable
 import org.apache.calcite.util.{Bug, BuiltInMethod, ImmutableBitSet, Util}
-
 import com.google.common.collect.ImmutableSet
-
 import java.util
 
+import org.apache.calcite.plan.hep.HepRelVertex
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala:47]: reorder import

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala [hunk 5]
```diff
@@ -244,7 +243,25 @@ class FlinkRelMdUniqueKeys private extends MetadataHandler[BuiltInMetadata.Uniqu
       ignoreNulls: Boolean): JSet[ImmutableBitSet] = {
     val inputUniqueKeys = mq.getUniqueKeys(rel.getInput, ignoreNulls)
     val rankFunColumnIndex = RankUtil.getRankNumberColumnIndex(rel).getOrElse(-1)
-    if (rankFunColumnIndex < 0) {
+    //TODO current deduplicate on row time is still a Rank,
+    // remove this after support deduplicate on row time
+    val canConvertToDeduplicate: Boolean = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala:248]: it's better we can put them into a separate commit

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala [hunk 6]
```diff
@@ -244,7 +243,25 @@ class FlinkRelMdUniqueKeys private extends MetadataHandler[BuiltInMetadata.Uniqu
...
+      val rankRange = rel.rankRange
+      val isRowNumberType = rel.rankType == RankType.ROW_NUMBER
+      val isLimit1 = rankRange match {
+        case rankRange: ConstantRankRange =>
+          rankRange.getRankStart() == 1 && rankRange.getRankEnd() == 1
+        case _ => false
+      }
+      !rel.outputRankNumber && isRowNumberType && isLimit1
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala:256]: `!rel.outputRankNumber` is unnecessary ? because even Rank output rank number, its value is always 1

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala [hunk 7]
```diff
@@ -244,7 +243,25 @@ class FlinkRelMdUniqueKeys private extends MetadataHandler[BuiltInMetadata.Uniqu
...
+    }
+
+    if (canConvertToDeduplicate) {
+      val retSet = new JHashSet[ImmutableBitSet]
+      retSet.add(rel.partitionKey)
+      retSet
+    }
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala:263]: please add related tests in FlinkRelMdUniqueKeysTest

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala [hunk 8]
```diff
@@ -553,6 +570,20 @@ class FlinkRelMdUniqueKeys private extends MetadataHandler[BuiltInMetadata.Uniqu
...
+      ignoreNulls: Boolean): JSet[ImmutableBitSet] = {
+    mq.getUniqueKeys(subset.getCurrentRel, ignoreNulls)
+  }
+
+  def getUniqueKeys(
+      subset: WatermarkAssigner,
+      mq: RelMetadataQuery,
+      ignoreNulls: Boolean): JSet[ImmutableBitSet] = {
+    mq.getUniqueKeys(subset.getInput, ignoreNulls)
+  }
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala:585]: ditto

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/TemporalJoinRewriteWithUniqueKeyRule.scala [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        .filter(_.nonEmpty)
+        .map(_.toArray
+          .map(fields)
+          .map(f => rexBuilder.makeInputRef(
+            f.getType,
+            leftFieldCnt + rightFields.indexOf(f)))
+          .toSeq)
+        .toArray
+        .sortBy(_.length)
+        .headOption
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/TemporalJoinRewriteWithUniqueKeyRule.scala:122]: it's hard to read, it's better to add some comments and split them into a few steps

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/TemporalJoinRewriteWithUniqueKeyRule.scala [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+      case t: TimeIndicatorRelDataType if t.isEventTime => true
+      case _ => false
+    }
+}
+
+object TemporalJoinRewriteWithUniqueKeyRule {
+  val INSTANCE: TemporalJoinRewriteWithUniqueKeyRule = new TemporalJoinRewriteWithUniqueKeyRule
+}
+
+
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/TemporalJoinRewriteWithUniqueKeyRule.scala:139]: nit: redundant blank

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/TemporalJoinUtil.scala [hunk 11]
```diff
@@ -17,75 +17,152 @@
...
-  }
 
   def makeRowTimeTemporalJoinConditionCall(
     rexBuilder: RexBuilder,
     leftTimeAttribute: RexNode,
     rightTimeAttribute: RexNode,
-    rightPrimaryKeyExpression: RexNode): RexNode = {
+    leftJoinKeyExpression: Seq[RexNode],
+    rightJoinKeyExpression: Seq[RexNode],
+    rightPrimaryKeyExpression: Seq[RexNode]): RexNode = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/TemporalJoinUtil.scala:108]: indent

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/TemporalJoinUtil.scala [hunk 12]
```diff
@@ -17,75 +17,152 @@
...
+      TEMPORAL_JOIN_RIGHT_KEY,
+      keyExpression)
+  }
+
   def makeProcTimeTemporalJoinConditionCall(
     rexBuilder: RexBuilder,
     leftTimeAttribute: RexNode,
-    rightPrimaryKeyExpression: RexNode): RexNode = {
+    leftJoinKeyExpression: Seq[RexNode],
+    rightJoinKeyExpression: Seq[RexNode]): RexNode = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/TemporalJoinUtil.scala:160]: indent

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/TemporalJoinUtil.scala [hunk 13]
```diff
@@ -103,4 +180,25 @@ object TemporalJoinUtil {
...
+  def isRowTimeJoin(rexBuilder: RexBuilder, joinInfo: JoinInfo): Boolean = {
+    val nonEquiJoinRex = joinInfo.getRemaining(rexBuilder)
+
+    var rowtimeJoin: Boolean = false
+    val visitor = new RexVisitorImpl[Unit](true) {
+      override def visitCall(call: RexCall): Unit = {
+        if (isRowTimeTemporalJoinConditionCall(call)) {
+           rowtimeJoin = true
+        } else {
+          call.getOperands.foreach(node => node.accept(this))
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/TemporalJoinUtil.scala:192]: super.visitCall(call)

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/join/TemporalJoinTest.scala [hunk 14]
```diff
@@ -17,95 +17,136 @@
...
+    " CREATE VIEW latestView as SELECT currency, rate, proctime FROM " +
+      "  (SELECT *, " +
+      "          ROW_NUMBER() OVER (PARTITION BY currency ORDER BY proctime DESC) AS rowNum " +
+      "   FROM RatesOnly" +
+      "  ) T" +
+      "  WHERE rowNum = 1")
+
+  util.addTable("CREATE VIEW latest_rates AS SELECT currency, LAST_VALUE(rate) AS rate " +
+    "FROM RatesHistory " +
+    "GROUP BY currency ")
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/join/TemporalJoinTest.scala:103]: move them into `before` method

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/join/TemporalJoinTest.scala [hunk 15]
```diff
@@ -17,129 +17,256 @@
...
+/**
+ * Test temporal join in stream mode.
+ */
 class TemporalJoinTest extends TableTestBase {
 
   val util: StreamTableTestUtil = streamTestUtil()
 
-  private val orders = util.addDataStream[(Long, String)](
-    "Orders", 'o_amount, 'o_currency, 'o_rowtime.rowtime)
+  util.addTable(
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/join/TemporalJoinTest.scala:32]: ditto

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/join/TemporalJoinTest.scala [hunk 16]
```diff
@@ -17,129 +17,256 @@
...
+    val sqlQuery4 = "SELECT " +
+      "o_amount * rate as rate " +
+      "FROM Orders AS o JOIN " +
+      "versionedTableWithoutRowtime FOR SYSTEM_TIME AS OF o.o_rowtime as r " +
+      "on o.o_currency = r.currency"
+    expectExceptionThrown(
+      sqlQuery4,
+      s"Event-Time Temporal Table Join requires both primary key and row time attribute in " +
+        s"versioned table, but no row time attribute found.",
+      classOf[ValidationException])
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/join/TemporalJoinTest.scala:243]: another case: join keys does not match pk

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/join/TemporalJoinTest.scala [hunk 17]
```diff
@@ -17,129 +17,256 @@
...
-      "SELECT * FROM " +
-        "(SELECT " +
-        "o_amount * rate as rate, " +
-        "secondary_key as secondary_key " +
-        "FROM Orders AS o, " +
-        "LATERAL TABLE (Rates(o_rowtime)) AS r " +
-        "WHERE currency = o_currency OR secondary_key = o_secondary_key), " +
-        "Table3 " +
-        "WHERE t3_secondary_key = secondary_key"
+  def testProcTimeTemporalJoinWithView(): Unit = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/join/TemporalJoinTest.scala:140]: other cases: non-equal join condition,  equal join condition with other predicates, etc

**Review by godfreyhe — CHANGES_REQUESTED** [review]
Thanks for the contribution @leonardBang , I left some comments. 
Btw, we should add unit tests for each added rule, you can refer to `LogicalUnnestRuleTest`

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRule.scala [hunk 18]
```diff
@@ -40,12 +51,105 @@ abstract class LogicalCorrelateToJoinFromTemporalTableRule(
...
+      case _ => node
+    }
+  }
+
+  protected def validateSnapshotInCorrelate(
+      snapshot: LogicalSnapshot,
+      correlate: LogicalCorrelate): Unit = {
+    // period specification check
+    snapshot.getPeriod.getType match {
+      // validate type is  event-time or processing time
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRule.scala:68]: this should ok that keeps line with following `case`

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRule.scala [hunk 19]
```diff
@@ -40,12 +51,105 @@ abstract class LogicalCorrelateToJoinFromTemporalTableRule(
...
+      case t: TimeIndicatorRelDataType =>
+      case _ =>
+        throw new ValidationException("Temporal table join currently only supports " +
+          "'FOR SYSTEM_TIME AS OF' left table's time attribute field")
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRule.scala:72]: `A JOIN B FOR SYSTEM_TIME AS OF A.proctime AS B_temporal` here we check wether user used the left time attribute column followed `AS OF` or not

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRule.scala [hunk 20]
```diff
@@ -74,23 +178,150 @@ abstract class LogicalCorrelateToJoinFromTemporalTableRule(
...
+    val snapshotTimeInputRef = extractSnapshotTimeInputRef(leftInput, snapshot)
+      .getOrElse(throw new ValidationException("Temporal Table Join requires time attribute in the " +
+        s"left table, but no row time attribute found."))
+
+    val rexBuilder = correlate.getCluster.getRexBuilder
+    val temporalCondition = if(isRowTimeTemporalTableJoin(snapshot)) {
+      val rightTimeInputRef = extractRightTimeInputRef(leftInput, snapshot)
+      if (rightTimeInputRef.isEmpty || !isRowtimeIndicatorType(rightTimeInputRef.get.getType)) {
+          throw new ValidationException("Event-Time Temporal Table Join requires both" +
+            s" primary key and row time attribute in versioned table," +
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRule.scala:277]: The `primary key` will checked in `TemporalJoinRewriteWithUniqueKeyRule.`

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecTemporalJoin.scala [hunk 21]
```diff
@@ -59,21 +64,21 @@ class StreamExecTemporalJoin(
...
-        } else {
-          call.getOperands.foreach(node => node.accept(this))
-        }
+  def rightInputUniqueKeyContainsJoinKey(): Boolean = {
+    val right = getInput(1)
+    val rightUniqueKeys = getCluster.getMetadataQuery.getUniqueKeys(right)
+    if (rightUniqueKeys != null) {
+      val joinKeys = keyPairs.map(_.target).toArray
+      rightUniqueKeys.exists {
+        uniqueKey => joinKeys.forall(uniqueKey.toArray.contains(_))
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecTemporalJoin.scala:73]: good catch

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala (hunk 5 above)
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala:248]: yes, this part will be rework in the commit of supporting deduplicate on row time.

**Comment by leonardBang** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRuleTest.scala [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+import org.apache.flink.api.scala._
+import org.apache.flink.table.api._
+import org.apache.flink.table.planner.plan.optimize.program.{FlinkChainedProgram, FlinkHepRuleSetProgramBuilder, HEP_RULES_EXECUTION_TYPE, StreamOptimizeContext}
+import org.apache.flink.table.planner.utils.{StreamTableTestUtil, TableTestBase, TableTestUtil}
+import org.junit.{Before, Test}
+
+/**
+ * Test for [[LogicalCorrelateToJoinFromTemporalTableRule]].
+ */
+class LogicalCorrelateToJoinFromTemporalTableRuleTest extends TableTestBase {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRuleTest.scala:34]: add a test about right join and full join

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRuleTest.scala [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+
+  @Test
+  def testRowTimeLeftTemporalJoinWithFilter(): Unit = {
+    setUpCurrentRule(LogicalCorrelateToJoinFromTemporalTableRule.WITH_FILTER)
+    util.verifyPlan("SELECT * FROM T1 LEFT JOIN T3 FOR SYSTEM_TIME AS OF T1.rowtime AS dimTable " +
+      "ON T1.id = dimTable.id AND dimTable.rate > 10")
+  }
+
+  @Test
+  def testLookupJoinOnTrue(): Unit = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/LogicalCorrelateToJoinFromTemporalTableRuleTest.scala:108]: also add a test about non-equal condition

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecTemporalJoin.scala [hunk 24]
```diff
@@ -241,7 +233,7 @@ class StreamExecTemporalJoinToCoProcessTranslator private (
     val minRetentionTime = tableConfig.getMinIdleStateRetentionTime
     val maxRetentionTime = tableConfig.getMaxIdleStateRetentionTime
     joinType match {
-      case JoinRelType.INNER =>
+      case JoinRelType.INNER | JoinRelType.LEFT =>
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecTemporalJoin.scala:236]: Does runtime support left join? this change should be covered by at least one IT case

Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecTemporalJoin.scala:236]: Yes, runtime will support left join in next PR

**Comment by leonardBang** [comment]
@flinkbot run azure

**Comment by leonardBang** [comment]
The planner tests passed, the failed e2e tests is not related to this PR, and it's caused by timeout. 
 cc: @godfreyhe  could you help to merge if you have no more concern?

**Comment by godfreyhe** [comment]
related timeout issue: https://issues.apache.org/jira/browse/FLINK-19664
